### PR TITLE
Make nix-shell evaluation faster

### PR DESCRIPTION
**Pull request by rvl**
### Issue Number

ADP-382 / #1888 

### Overview

- Updates Haskell.nix rev to latest.
- Use the "materialization" feature so that evaluation of the nix-shell (done locally and on hydra) is cached as much as possible.

### Comments

```
$ GC_PRINT_STATS=1 command time nix-shell --option substitute false --run "cabal --version"

master branch:
  ...
  Grow heap to 2786488 KiB after 806218912 bytes allocated
  cabal-install version 3.2.0.0
  compiled using version 3.2.0.0 of the Cabal library 
  30.51user 2.34system 0:29.05elapsed 113%CPU (0avgtext+0avgdata 3300584maxresident)k
  0inputs+0outputs (0major+915357minor)pagefaults 0swaps

PR branch:
  ...
  Grow heap to 2393272 KiB after 442027296 bytes allocated
  cabal-install version 3.2.0.0
  compiled using version 3.2.0.0 of the Cabal library 
  26.63user 1.84system 0:24.78elapsed 114%CPU (0avgtext+0avgdata 2889004maxresident)k
  0inputs+0outputs (0major+812718minor)pagefaults 0swaps

memory usage: 2.3GiB  ( -14%)
startup time: 0:24.78 ( -15%)
```

### CODE REVIEW

File: shell.nix [hunk 1]
```diff
@@ -1,16 +1,117 @@
...
+      #   nix-shell --arg checkMaterialization true
+      index-state = "2020-10-20T00:00:00Z";
+      mkTool = name: args: {
+        inherit index-state checkMaterialization;
+      } // (if checkMaterialization then builtins.removeAttrs args ["plan-sha256"] else args);
+    in mapAttrs mkTool {
+      cabal = {
+        version = "3.2.0.0";
+        plan-sha256 = "0ik4ws852dk7wchbhc84w5ac149myvc1az7rxapfy0hdmiiwjh2j
+";
```
Comment (hamishmack) [inline shell.nix]: Ooops

**Comment by hamishmack** [comment]
I am not sure if we still need `plan-sha256` values.  You might want to try without them as it would probably make updating the materialized nix easier.

**Review by hamishmack — APPROVED** [review]

File: shell.nix (hunk 1 above)
Comment (rvl) [reply shell.nix]: Oops, fixed.

**Comment by rvl** [comment]
@hamishmack I tried removing the `plan-sha256` lines but was then getting these kind of messages from the nix-shell:
```
building '/nix/store/hq5bl3bc3f6cps36yaw28dy89czgz68l-hoogle-plan-to-nix-pkgs.drv'...
```
So perhaps they do help with caching the cabal plan.

**Comment by hamishmack** [comment]
Is that derivation actually running `cabal configure` though or just copying the files from the `materialized` folder.  Perhaps check `nix show-derivation /nix/store/hq5bl3bc3f6cps36yaw28dy89czgz68l-hoogle-plan-to-nix-pkgs.drv`.  I think it should only be one of these https://github.com/input-output-hk/haskell.nix/blob/b06fdfa06595668b367a8b9298792076c87e4618/lib/materialize.nix#L116-L120.

**Comment by rvl** [comment]
I see, yes you are right, it was just copying files.
I have removed plan-sha256 and cleaned up the update instructions.

**Pull request merged**